src/main/java/org/opensearch/neuralsearch/search/MultiLeafFieldComparator.java

```diff
@@ -0,0 +1,95 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package org.opensearch.neuralsearch.search;
+
+import java.io.IOException;
+import org.apache.lucene.search.DocIdSetIterator;
+import org.apache.lucene.search.LeafFieldComparator;
+import org.apache.lucene.search.Scorable;
+
+final class MultiLeafFieldComparator implements LeafFieldComparator {
+
+    private final LeafFieldComparator[] comparators;
+    private final int[] reverseMul;
+    // we extract the first comparator to avoid array access in the common case
+    // that the first comparator compares worse than the bottom entry in the queue
+    private final LeafFieldComparator firstComparator;
+    private final int firstReverseMul;
+
+    MultiLeafFieldComparator(LeafFieldComparator[] comparators, int[] reverseMul) {
+        if (comparators.length != reverseMul.length) {
+            throw new IllegalArgumentException(
+                "Must have the same number of comparators and reverseMul, got " + comparators.length + " and " + reverseMul.length
```

Comment:
please use proper formatting for complex error message - String.format + placeholders for parameters